contracts/upgradeable_contracts/ChaiConnector.sol

```diff
@@ -0,0 +1,269 @@
+    */
+    function removeChaiToken() external onlyOwner chaiTokenEnabled {
+        _convertChaiToDai(investedAmountInDai());
+        _payInterest();
+        delete boolStorage[CHAI_TOKEN_ENABLED];
+    }
+
+    /**
+     * @return Configured address of a receiver
+     */
+    function interestReceiver() public view returns (ERC677Receiver) {
+        return ERC677Receiver(addressStorage[INTEREST_RECEIVER]);
+    }
+
+    /**
+     * Updates interest receiver contract address
+     * @param receiver New receiver contract address
+     */
+    function setInterestReceiver(address receiver) external onlyOwner {
+        addressStorage[INTEREST_RECEIVER] = receiver;
+    }
+
+    /**
+     * @return Timestamp of last interest payment
+     */
+    function lastInterestPayment() public view returns (uint256) {
+        return uintStorage[LAST_TIME_INTEREST_PAYED];
+    }
+
+    /**
+     * @return Configured minimum interest collection period
+     */
+    function interestCollectionPeriod() public view returns (uint256) {
+        return uintStorage[INTEREST_COLLECTION_PERIOD];
+    }
+
+    /**
+     * @dev Configures minimum interest collection period
+     * @param period collection period
+     */
+    function setInterestCollectionPeriod(uint256 period) external onlyOwner {
+        uintStorage[INTEREST_COLLECTION_PERIOD] = period;
+    }
+
+    /**
+    * @dev Pays all available interest, in Dai tokens.
+    * Upgradeability owner can call this method without time restrictions,
+    * for others, the method can be called only once a specified period.
+    */
+    function payInterest() external chaiTokenEnabled {
+        // solhint-disable not-rely-on-time
+        if (
+            lastInterestPayment() + interestCollectionPeriod() < now ||
+            IUpgradeabilityOwnerStorage(this).upgradeabilityOwner() == msg.sender
+        ) {
+            uintStorage[LAST_TIME_INTEREST_PAYED] = now;
+            _payInterest();
+        }
+        // solhint-enable not-rely-on-time
+    }
+
+    /**
+    * @dev Internal function for paying all available interest, in Dai tokens
+    */
+    function _payInterest() internal {
+        require(address(interestReceiver()) != address(0));
+
+        // since investedAmountInChai() returns a ceiled value,
+        // the value of chaiBalance() - investedAmountInChai() will be floored,
+        // leading to excess remaining chai balance
+        uint256 balanceBefore = daiBalance();
+        chaiToken().exit(address(this), chaiBalance().sub(investedAmountInChai()));
+        uint256 interestInDai = daiBalance().sub(balanceBefore);
+
+        erc20token().transfer(interestReceiver(), interestInDai);
+
+        if (AddressUtils.isContract(interestReceiver())) {
+            interestReceiver().onTokenTransfer(address(this), interestInDai, "");
+        }
+
+        require(dsrBalance() >= investedAmountInDai());
```

Comment:
What is purpose of this check? To rollback the transaction if something has gone wrong? Does it make sense to introduce similar check into `convertDaiToChai` and `_ convertChaiToDai`?

Reply:
I didn't put such check in convert operations, since there are not changing the total balance of a bridge. On the other hand, `payInterest` decreases the overall balance of the bridge, so I have added this extra check.

Reply:
So, from your perspective, what is better to do with this checks?

Reply:
I would add such check to `convertDaiToChai` just to make sure that the Chai/DSR/whatever contract which we do not control has not started to behave unexpectedly. The situation when we cannot convert Dai to Chai due to reverted transaction is better than unexpected behavior leading to unbalancing of the bridge. 

I would add such check to `_convertChaiToDai` since it will lead to failure of `executeSignatures` call. Failure of the call will bring our attention on the problem. So, if due to some reason (probably a bug) Chai/DSR/whatever will not convert the tokens properly we will notice and will try to handle this situation. Otherwise it could lead to unbalancing of the bridge as well.

What do you think?

Reply:
I think, it is not a problem to have such checks. So let's add them.

In case of `_convertChaiToDai` it is not a problem at all. However, there is a problematic moment in `convertDaiToChai`. Let me explain:

When evaluating the amount of DAI kept in Chai using `dsrBalance()`, there are some fixed point truncations. 
The dependency between invested amount of DAI - `value` and returned value of `dsrBalance()` - `res` is the following:
`res = floor(floor(value / chi) * chi))`, 
where `chi` is a fixed point number with precision of 27 digits. This coefficient is increasing every second, currently it is ~ `1.012`.
This, if I am not mistaken, can lead up to losses of `ceil(chi)` DAI in this balance evaluation.
 
This is not a problem for the bridge itself, since after even one block, the collected interest will be orders more than this error.
But, this reasoning implies that the check should be different:
`require(dsrBalance() >= investedAmountInDai() - (1 + chaiToken().pot().chi() / 10**27));`

Reply:
Also, I am not quite sure if this will work properly for several calls to `convertDaiToChai()` in the same block. I will try to test such scenario, to figure that out.

Reply:
One more option for this check, is to use high enough constant in inequality (e. g. `require(dsrBalance() >= investedAmountInDai() - 1000000)`)
Such approach will definitely work for all cases.

Reply:
I am OK with any option especially if it is well documented in the code.

Reply:
Done